readthedocs/oauth/services/githubapp.py

```diff
@@ -0,0 +1,473 @@
+        """
+        has_error = False
+        # Refresh access to all installations accessible to the user.
+        for service in cls.for_user(user):
+            try:
+                service.sync()
+            except SyncServiceError:
+                # Don't stop the sync if one installation fails,
+                # as we should try to sync all installations.
+                has_error = True
+
+        # Update the access to each repository the user has access to.
+        queryset = RemoteRepository.objects.filter(
+            remote_repository_relations__user=user,
+            vcs_provider=cls.vcs_provider_slug,
+        )
+        for repository in queryset:
+            service = cls(repository.github_app_installation)
+            service.update_or_create_repositories([int(repository.remote_id)])
+
+        # Update access to each organization the user has access to.
+        queryset = RemoteOrganization.objects.filter(
+            remote_organization_relations__user=user,
+            vcs_provider=cls.vcs_provider_slug,
+        )
+        for remote_organization in queryset:
+            remote_repo = remote_organization.repositories.first()
+            # NOTE: this should never happen, unless our data is out of sync
+            # (we delete orphaned organizations when deleting projects).
+            if not remote_repo:
+                log.info(
+                    "Remote organization without repositories detected, deleting.",
+                    organization_login=remote_organization.slug,
+                    remote_id=remote_organization.remote_id,
+                )
+                remote_organization.delete()
+                continue
+
+            service = cls(remote_repo.github_app_installation)
+            service.update_or_create_organization(remote_organization.slug)
+
+        if has_error:
+            raise SyncServiceError()
+
+    def sync(self):
+        """
+        Sync all repositories and organizations that are accessible to the installation.
+
+        Repositories that are no longer accessible to the installation are removed from the database
+        only if they are not linked to a project. This is in case the user wants to grant access to the repository again.
+
+        If a remote organization doesn't have any repositories after removing the repositories,
+        we remove the organization from the database.
+        """
+        try:
+            app_installation = self.app_installation
+        except GithubException as e:
+            log.info(
+                "Failed to get installation",
+                installation_id=self.installation.installation_id,
+                exc_info=True,
+            )
+            if e.status == 404:
+                # The app was uninstalled, we remove the installation from the database.
+                self.installation.delete()
```

Comment:
This seems aggressive -- are we sure this is always the required outcome with a 404?

Reply:
There is no other way to know if we have access to the installation.

Reply:
Nope, what other reason would be for GH to return a 404 other than losing access to the installation? GH could return 401/5xx and others, that's why it specifically checks for a 404.